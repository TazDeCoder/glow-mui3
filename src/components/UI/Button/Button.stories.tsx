import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { Button } from './Button';
import { Box } from '../../Utils';

export default {
  title: 'Components/UI/Button',
  component: Button,
  argTypes: {
    onClick: { action: 'clicked' },
  },
} as ComponentMeta<typeof Button>;

export const Default: ComponentStory<typeof Button> = (args) => (
  <Button {...args}>Hello, World!</Button>
);

export const Colors: ComponentStory<typeof Button> = (args) => (
  <Box sx={{ display: 'flex', columnGap: 2 }}>
    <Button {...args} color="primary">
      Primary
    </Button>
    <Button {...args} color="secondary">
      Secondary
    </Button>
    <Button {...args} color="info">
      Info
    </Button>
    <Button {...args} color="error">
      Error
    </Button>
    <Button {...args} color="warning">
      Warning
    </Button>
    <Button {...args} color="success">
      Success
    </Button>
  </Box>
);

export const Variants: ComponentStory<typeof Button> = (args) => (
  <Box sx={{ display: 'flex', columnGap: 2 }}>
    <Button {...args} variant="filled">
      Filled
    </Button>
    <Button {...args} variant="outlined">
      Outlined
    </Button>
    <Button {...args} variant="text">
      Text
    </Button>
  </Box>
);
