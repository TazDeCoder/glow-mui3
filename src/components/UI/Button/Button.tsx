/** @jsxRuntime classic */
/** @jsx jsx */

import { forwardRef } from 'react';
import { jsx, ThemeUICSSObject, ThemeUIStyleObject } from '@theme-ui/core';

import { Box as _Box, BoxProps, BoxOwnProps } from '../../Utils';
import { Typography } from '../Typography';
import { __internalProps } from '../../shared/utils';
import type { Assign, ForwardRef, ColorTheme } from '../../types';

const Box = _Box as React.ForwardRefExoticComponent<
BoxProps & React.RefAttributes<HTMLButtonElement>
>;

export interface ButtonProps
  extends Assign<React.ComponentPropsWithRef<'button'>, BoxOwnProps> {
  /** Color theme used for the component
   * @default 'primary'
   */
  color?: ColorTheme;
  /** Variant style of the component
   * @default 'filled'
   */
  variant?: 'filled' | 'outlined' | 'text';
  /** Disabled state of the component
   * @default false
   */
  disabled?: boolean;
}

const __css: ThemeUICSSObject = {
  position: 'relative',
  minWidth: 48,
  height: 40,
  px: 24,
  overflow: 'hidden',
  border: '1px solid currentColor',
  borderRadius: 'full',
  transition: 'standard',
  '&:hover::after': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    bg: 'action.hover',
  },
  '&:focus::after': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    bg: 'action.active',
  },
};

export const Button: ForwardRef<HTMLButtonElement, ButtonProps> = forwardRef(
  (
    {
      children,
      color = 'primary',
      variant = 'filled',
      type = 'button',
      disabled = false,
      sx,
      ...props
    }: ButtonProps,
    ref,
  ) => {
    const style: ThemeUIStyleObject = {
      ...(!disabled
        ? {
          cursor: 'pointer',
          ...(variant === 'filled' && {
            color: `${color}.container`,
            bg: `${color}.main`,
            transition: 'emphasized',
            '&:hover': {
              boxShadow: 'sm',
            },
          }),
          ...(variant === 'outlined' && {
            color: `${color}.main`,
            bg: 'transparent',
            borderColor: 'outline.main',
            '&:focus': {
              borderColor: `${color}.main`,
            },
            '&:active': {
              borderColor: 'outline.main',
            },
          }),
          ...(variant === 'text' && {
            bg: 'transparent',
            borderColor: 'transparent',
          }),
        }
        : {
          opacity: 'disabled',
          color: 'surface.onMain',
          bg: 'transparent',
          borderColor:
              variant === 'outlined' ? 'surface.onMain' : 'transparent',
          userSelect: 'none',
          '&::before': {
            content: '""',
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            opacity: 0.12,
            bg: variant === 'filled' && 'surface.onMain',
          },
          '&:hover::after, &:focus::after': {
            content: 'none',
          },
        }),
    };

    return (
      <Box
        ref={ref}
        as="button"
        sx={{ ...style, ...sx }}
        {...({ type, disabled, ...props } as BoxProps)}
        {...__internalProps({ __css })}
      >
        <Typography
          variant="label"
          color={
            disabled
              ? 'surface.onMain'
              : variant === 'filled'
                ? `${color}.onMain`
                : `${color}.main`
          }
        >
          {children}
        </Typography>
      </Box>
    );
  },
);
