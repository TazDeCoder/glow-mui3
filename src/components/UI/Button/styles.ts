import type { ThemeUIStyleObject } from '@theme-ui/core';

export const baseStyles: ThemeUIStyleObject = {
  position: 'relative',
  minWidth: 48,
  height: 40,
  px: 24,
  border: '1px solid',
  borderRadius: 'full',
  cursor: 'pointer',
};

export const interactionStyles: ThemeUIStyleObject = {
  '&:hover::after': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    bg: 'action.hover',
    borderRadius: 'inherit',
  },
  '&:focus::after': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    bg: 'action.active',
    borderRadius: 'inherit',
  },
};

export const filledStyles: ThemeUIStyleObject = {
  transition: 'emphasized',
  '&:hover': {
    boxShadow: 'sm',
  },
};

export const outlinedStyles: ThemeUIStyleObject = {
  bg: 'transparent',
  borderColor: 'outline.main',
  transition: 'standard',
  '&:active': {
    borderColor: 'outline.main',
  },
};

export const textStyles: ThemeUIStyleObject = {
  bg: 'transparent',
  borderColor: 'transparent',
};
