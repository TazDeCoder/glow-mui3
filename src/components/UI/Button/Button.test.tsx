import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Button } from './Button';

describe('<Button />', () => {
  describe('Snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(<Button />);
      expect(container).toMatchSnapshot();
    });
  });
  describe('User events', () => {
    it('triggers onClick handler when clicked', async () => {
      const clickHandler = jest.fn();
      const { getByRole } = render(
        <Button onClick={clickHandler}>Accept</Button>,
      );
      const btnEl = getByRole('button');
      await userEvent.click(btnEl);
      expect(clickHandler).toHaveBeenCalled();
    });
  });
});
