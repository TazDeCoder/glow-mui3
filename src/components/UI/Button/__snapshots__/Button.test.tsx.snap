// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Snapshots renders correctly 1`] = `
.emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  min-width: 48px;
  height: 40px;
  padding-left: 24px;
  padding-right: 24px;
  overflow: hidden;
  border: 1px solid currentColor;
  border-radius: full;
  -webkit-transition: standard;
  transition: standard;
  cursor: pointer;
  color: primary.container;
  background-color: primary.main;
  -webkit-transition: emphasized;
  transition: emphasized;
}

.emotion-0:hover::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: action.hover;
}

.emotion-0:focus::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: action.active;
}

.emotion-0:hover {
  box-shadow: sm;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: primary.onMain;
}

<div>
  <button
    class="emotion-0"
    type="button"
  >
    <span
      class="emotion-1"
    />
  </button>
</div>
`;
